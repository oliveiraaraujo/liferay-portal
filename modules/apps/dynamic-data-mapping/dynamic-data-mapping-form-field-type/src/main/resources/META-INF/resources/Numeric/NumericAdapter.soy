{namespace NumericAdapter}

/**
 * NumericAdapter
 */
{template .render}
	<div key="ddm-field-type-numeric"></div>
{/template}