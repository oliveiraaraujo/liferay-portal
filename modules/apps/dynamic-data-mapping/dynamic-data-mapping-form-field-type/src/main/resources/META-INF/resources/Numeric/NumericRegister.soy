{namespace NumericRegister}

{deltemplate PageRenderer.RegisterFieldType variant="'numeric'"}
	{call NumericAdapter.render data="all" /}
{/deltemplate}