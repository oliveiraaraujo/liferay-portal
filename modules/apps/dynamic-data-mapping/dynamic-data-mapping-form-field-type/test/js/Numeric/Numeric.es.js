/**
 * Copyright (c) 2000-present Liferay, Inc. All rights reserved.
 *
 * This library is free software; you can redistribute it and/or modify it under
 * the terms of the GNU Lesser General Public License as published by the Free
 * Software Foundation; either version 2.1 of the License, or (at your option)
 * any later version.
 *
 * This library is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
 * FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public License for more
 * details.
 */

import {fireEvent} from '@testing-library/react';

import Numeric from '../../../src/main/resources/META-INF/resources/Numeric/Numeric.es';
import withContextMock from '../__mocks__/withContextMock.es';

let component;
const spritemap = 'icons.svg';

const defaultNumericConfig = {
	name: 'numericField',
	spritemap,
};

const NumericWithContextMock = withContextMock(Numeric);

describe('Field Numeric', () => {
	beforeEach(() => {
		jest.useFakeTimers();
	});

	afterEach(() => {
		if (component) {
			component.dispose();
		}
	});

	it('renders the default markup', () => {
		component = new NumericWithContextMock({
			...defaultNumericConfig,
			readOnly: false,
		});

		expect(component).toMatchSnapshot();
	});

	it('is not readOnly', () => {
		component = new NumericWithContextMock({
			...defaultNumericConfig,
			readOnly: false,
		});

		expect(component).toMatchSnapshot();
	});

	it('has a helptext', () => {
		component = new NumericWithContextMock({
			...defaultNumericConfig,
			tip: 'Type something',
		});

		expect(component).toMatchSnapshot();
	});

	it('has an id', () => {
		component = new NumericWithContextMock({
			...defaultNumericConfig,
			id: 'ID',
		});

		expect(component).toMatchSnapshot();
	});

	it('has a label', () => {
		component = new NumericWithContextMock({
			...defaultNumericConfig,
			label: 'label',
		});

		expect(component).toMatchSnapshot();
	});

	it('has a placeholder', () => {
		component = new NumericWithContextMock({
			...defaultNumericConfig,
			placeholder: 'Placeholder',
		});

		expect(component).toMatchSnapshot();
	});

	it('is not required', () => {
		component = new NumericWithContextMock({
			...defaultNumericConfig,
			required: false,
		});

		expect(component).toMatchSnapshot();
	});

	it('renders Label if showLabel is true', () => {
		component = new NumericWithContextMock({
			...defaultNumericConfig,
			label: 'Numeric Field',
			showLabel: true,
		});

		expect(component).toMatchSnapshot();
	});

	it('has a spritemap', () => {
		component = new NumericWithContextMock(defaultNumericConfig);

		expect(component).toMatchSnapshot();
	});

	it('has a value', () => {
		component = new NumericWithContextMock({
			...defaultNumericConfig,
			value: '123',
		});

		expect(component).toMatchSnapshot();
	});

	it('has a key', () => {
		component = new NumericWithContextMock({
			...defaultNumericConfig,
			key: 'key',
		});

		expect(component).toMatchSnapshot();
	});

	it('propagates the field edit event', () => {
		jest.useFakeTimers();

		const fn = jest.fn();

		component = new NumericWithContextMock({
			...defaultNumericConfig,
			events: {fieldEdited: fn},
			key: 'input',
		});

		const input = component.element.querySelector('input');

		fireEvent.change(input, {target: {value: '2'}});

		jest.runAllTimers();

		expect(fn).toHaveBeenCalled();
	});

	it('changes the mask type', () => {
		component = new NumericWithContextMock({
			...defaultNumericConfig,
		});

		jest.runAllTimers();

		component.props.dataType = 'double';

		jest.runAllTimers();

		expect(component.props.dataType).toBe('double');
	});

	/**
	 * This test needs to be revised, it is strange to wait for the
	 * decimal value when the dataType is an integer.
	 */
	it.skip('check if event is sent when decimal is being writen', done => {
		const handleFieldEdited = data => {
			expect(data).toEqual(
				expect.objectContaining({
					fieldInstance: component,
					originalEvent: expect.any(Object),
					value: '3.0',
				})
			);
			done();
		};

		const events = {fieldEdited: handleFieldEdited};

		component = new NumericWithContextMock({
			...defaultNumericConfig,
			events,
			key: 'input',
		});

		const input = component.element.querySelector('input');

		fireEvent.change(input, {
			target: {
				value: '3.0',
			},
		});

		jest.runAllTimers();
	});

	it('check field value is rounded when fieldType is integer but it receives a double', done => {
		const handleFieldEdited = event => {
			expect(event.value).toBe('4');
			done();
		}

		const events = {fieldEdited: handleFieldEdited}

		component = new NumericWithContextMock({
			...defaultNumericConfig,
			events,
			key: 'input',
			value: '3.8',
		});

		jest.runAllTimers();
	});
});
