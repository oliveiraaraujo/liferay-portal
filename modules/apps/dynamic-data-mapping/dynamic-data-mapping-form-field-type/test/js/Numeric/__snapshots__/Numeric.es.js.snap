// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Field Numeric has a helptext 1`] = `
<div class="react-component-adapter">
  <div class="form-group" data-field-name="numericField"><input name="numericField" aria-label="numeric" type="text" class="form-control" value=""><span class="form-text">Type something</span></div>
</div>
`;

exports[`Field Numeric has a key 1`] = `
<div class="react-component-adapter">
  <div class="form-group" data-field-name="numericField"><input name="numericField" aria-label="numeric" type="text" class="form-control" value=""></div>
</div>
`;

exports[`Field Numeric has a label 1`] = `
<div class="react-component-adapter">
  <div class="form-group" data-field-name="numericField">
    <p class="ddm-label">label</p><input name="numericField" aria-label="numeric" type="text" class="form-control" value="">
  </div>
</div>
`;

exports[`Field Numeric has a placeholder 1`] = `
<div class="react-component-adapter">
  <div class="form-group" data-field-name="numericField"><input name="numericField" placeholder="Placeholder" aria-label="numeric" type="text" class="form-control" value=""></div>
</div>
`;

exports[`Field Numeric has a spritemap 1`] = `
<div class="react-component-adapter">
  <div class="form-group" data-field-name="numericField"><input name="numericField" aria-label="numeric" type="text" class="form-control" value=""></div>
</div>
`;

exports[`Field Numeric has a value 1`] = `
<div class="react-component-adapter">
  <div class="form-group" data-field-name="numericField"><input name="numericField" aria-label="numeric" type="text" class="form-control" value="123"></div>
</div>
`;

exports[`Field Numeric has an id 1`] = `
<div class="react-component-adapter">
  <div class="form-group" data-field-name="numericField"><input id="ID" name="numericField" aria-label="numeric" type="text" class="form-control" value=""></div>
</div>
`;

exports[`Field Numeric is not readOnly 1`] = `
<div class="react-component-adapter">
  <div class="form-group" data-field-name="numericField"><input name="numericField" aria-label="numeric" type="text" class="form-control" value=""></div>
</div>
`;

exports[`Field Numeric is not required 1`] = `
<div class="react-component-adapter">
  <div class="form-group" data-field-name="numericField"><input name="numericField" aria-label="numeric" type="text" class="form-control" value=""></div>
</div>
`;

exports[`Field Numeric renders Label if showLabel is true 1`] = `
<div class="react-component-adapter">
  <div class="form-group" data-field-name="numericField">
    <p class="ddm-label">Numeric Field</p><input name="numericField" aria-label="numeric" type="text" class="form-control" value="">
  </div>
</div>
`;

exports[`Field Numeric renders the default markup 1`] = `
<div class="react-component-adapter">
  <div class="form-group" data-field-name="numericField"><input name="numericField" aria-label="numeric" type="text" class="form-control" value=""></div>
</div>
`;
